cli/support.go

```diff
@@ -0,0 +1,209 @@
+package cli
+
+import (
+	"archive/zip"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"text/tabwriter"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/sloghuman"
+	"github.com/coder/coder/v2/cli/clibase"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/support"
+)
+
+func (r *RootCmd) support() *clibase.Cmd {
+	supportCmd := &clibase.Cmd{
+		Use:   "support { dump }",
+		Short: "Commands for troubleshooting issues with a Coder deployment.",
+		Handler: func(inv *clibase.Invocation) error {
+			return inv.Command.HelpHandler(inv)
+		},
+		Hidden: true, // TODO: un-hide once the must-haves from #12160 are completed.
+		Children: []*clibase.Cmd{
+			r.supportBundle(),
+		},
+	}
+	return supportCmd
+}
+
+func (r *RootCmd) supportBundle() *clibase.Cmd {
+	var outputPath string
+	client := new(codersdk.Client)
+	cmd := &clibase.Cmd{
+		Use:   "bundle <workspace> [<agent>]",
+		Short: "Generate a support bundle to troubleshoot issues connecting to a workspace.",
+		Long:  `This command generates a file containing detailed troubleshooting information about the Coder deployment and workspace connections. You must specify a single workspace (and optionally an agent name).`,
+		Middleware: clibase.Chain(
+			clibase.RequireRangeArgs(0, 2),
+			r.InitClient(client),
+		),
+		Handler: func(inv *clibase.Invocation) error {
+			var (
+				log = slog.Make(sloghuman.Sink(inv.Stderr)).
+					Leveled(slog.LevelDebug)
+				deps = support.Deps{
+					Client: client,
+					Log:    log,
+				}
+			)
+
+			if len(inv.Args) == 0 {
+				return xerrors.Errorf("must specify workspace name")
+			}
+			ws, err := namedWorkspace(inv.Context(), client, inv.Args[0])
+			if err != nil {
+				return xerrors.Errorf("invalid workspace: %w", err)
+			}
+
+			deps.WorkspaceID = ws.ID
+
+			agentName := ""
+			if len(inv.Args) > 1 {
+				agentName = inv.Args[1]
+			}
+
+			agt, found := findAgent(agentName, ws.LatestBuild.Resources)
+			if !found {
+				return xerrors.Errorf("could not find agent named %q for workspace", agentName)
+			}
+
+			deps.AgentID = agt.ID
+
+			if outputPath == "" {
```

Comment:
Could do this even before we fetch the workspace from API since it's not used as part of the filename.

Reply:
I think it makes more sense to do the up-front checks first; if the workspace doesn't even exist then there's no point in writing the file.